File: webui/hpm/js/term.js
// Copyright 2015 Eryx <evorui аt gmаil dοt cοm>, All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

var hpTerm = {
  taxonomy_ls_cache: null,
};

hpTerm.SpecActive = function () {
  return lynkui.storage.get("hpm_spec_active");
};

hpTerm.SpecTermModelActive = function (value) {
  if (!hpTerm.SpecActive()) {
    return null;
  }
  var k = "hpm_stm_" + hpTerm.SpecActive();
  if (value && value.length > 1) {
    lynkui.storage.set(k, value);
  }
  return lynkui.storage.get(k);
};

hpTerm.List = function (modname, modelid) {
  var alertid = "#hpm-node-alert",
    page = 0;

  if (!modname && hpTerm.SpecActive()) {
    modname = hpTerm.SpecActive();
  }
  if (!modelid && hpTerm.SpecTermModelActive()) {
    modelid = hpTerm.SpecTermModelActive();
  }
  if (lynkui.storage.get("hpm_termls_page")) {
    page = lynkui.storage.get("hpm_termls_page");
  }

  if (!modname || !modelid) {
    return;
  }

  var uri = "modname=" + modname + "&modelid=" + modelid + "&page=" + page;
  if (document.getElementById("qry_text")) {
    uri += "&qry_text=" + $("#qry_text").val();
  }

  var ep = lynkui.newEventProxy("tpl", "data", function (tpl, rsj) {
    if (tpl) {
      $("#work-content").html(tpl);
    }

    hpTerm.SpecTermModelActive(modelid);

    if (!rsj || rsj.kind != "TermList" || !rsj.items || rsj.items.length < 1) {
      $("#hpm-nodels").empty();
      $("#hpm-termls").empty();

      lynkui.alert.innerShow(alertid, "info", "Item Not Found");
    } else {
      $(alertid).hide();
    }
    $("#hpm-term-list-new-title").text("New " + rsj.model.title);

    if (!rsj.items) {
      rsj.items = [];
    }

    for (var i in rsj.items) {
      rsj.items[i].created = lynkui.utilx.unixTimeFormat(
        rsj.items[i].created,
        "Y-m-d"
      );
      rsj.items[i].updated = lynkui.utilx.unixTimeFormat(
        rsj.items[i].updated,
        "Y-m-d H:i:s"
      );

      if (!rsj.items[i].weight) {
        rsj.items[i].weight = 0;
      }

      if (!rsj.items[i].pid) {
        rsj.items[i].pid = 0;
      }

      if (rsj.model.type == "taxonomy" && rsj.items[i].pid == 0) {
        rsj.items[i]._subs = hpTerm.ListSubRange(
          rsj.items,
          null,
          rsj.items[i].id,
          0
        );
      }
    }

    hpTerm.taxonomy_ls_cache = rsj;

    lynkui.template.render({
      dstid: "hpm-termls",
      tplid: "hpm-termls-tpl",
      data: {
        model: rsj.model,
        modname: modname,
        modelid: modelid,
        items: rsj.items,
      },
      callback: function () {
        if (rsj.model.type != "taxonomy") {
          rsj.meta.RangeLen = 20;

          lynkui.template.render({
            dstid: "hpm-termls-pager",
            tplid: "hpm-termls-pager-tpl",
            data: hpMgr.Pager(rsj.meta),
          });
        } else {
          $("#hpm-termls-pager").empty();
        }

        hpNode.OpToolsRefresh("#hpm-node-term-opts");
      },
    });
  });

  ep.fail(function (err) {
    // TODO
    alert("SpecListRefresh error, Please try again later (EC:app-termlist)");
  });

  // template
  var el = document.getElementById("hpm-termls");
  if (!el || el.length < 1) {
    hpMgr.TplCmd("term/list", {
      callback: function (err, tpl) {
        if (err) {
          return ep.emit("error", err);
        }

        ep.emit("tpl", tpl);
      },
    });
  } else {
    ep.emit("tpl", null);
  }

  hpMgr.ApiCmd("term/list?" + uri, {
    callback: ep.done("data"),
  });
};

hpTerm.Sprint = function (num) {
  var s = "";
  for (i = 0; i < num; i++) {
    s += "&nbsp;&nbsp;&nbsp;&nbsp;";
  }

  return s;
};

hpTerm.ListSubRange = function (ls, rs, pid, dpnum) {
  if (!rs) {
    rs = [];
  }

  dpnum++;

  for (var i in rs) {
    if (rs[i].id == pid) {
      // return rs;
    }
  }

  for (var i in ls) {
    if (ls[i].pid == pid) {
      ls[i]._dp = dpnum;
      rs.push(ls[i]);
      rs = hpTerm.ListSubRange(ls, rs, ls[i].id, dpnum);
    }
  }

  return rs;
};

hpTerm.ListPage = function (page) {
  lynkui.storage.set("hpm_termls_page", parseInt(page));
  hpTerm.List();
};

hpTerm.Set = function (modname, modelid, termid) {
  var alertid = "#hpm-node-alert";

  if (!modname && hpTerm.SpecActive()) {
    modname = hpTerm.SpecActive();
  }
  if (!modelid && hpTerm.SpecTermModelActive()) {
    modelid = hpTerm.SpecTermModelActive();
  }

  if (!modname || !modelid) {
    return;
  }

  var uri = "modname=" + modname + "&modelid=" + modelid;

  var ep = lynkui.newEventProxy("tpl", "data", function (tpl, data) {
    if (!tpl) {
      return; // TODO
    }

    $("#work-content").html(tpl);

    if (!data || data.kind != "Term") {
      return lynkui.alert.innerShow(alertid, "info", "Item Not Found");
    }

    if (!data.status) {
      data.status = 1;
    }
    if (!data.weight) {
      data.weight = 0;
    }
    if (!data.pid) {
      data.pid = 0;
    }

    data._taxonomy_ls = hpTerm.taxonomy_ls_cache;

    $(alertid).hide();
    hpNode.OpToolsRefresh("clean");

    lynkui.template.render({
      dstid: "hpm-termset",
      tplid: "hpm-termset-tpl",
      data: data,
    });
  });

  ep.fail(function (err) {
    // TODO
    alert("SpecListRefresh error, Please try again later (EC:app-nodelist)");
  });

  hpMgr.TplCmd("term/set", {
    callback: function (err, tpl) {
      if (err) {
        return ep.emit("error", err);
      }
      ep.emit("tpl", tpl);
    },
  });

  if (termid) {
    hpMgr.ApiCmd("term/entry?" + uri + "&id=" + termid, {
      callback: ep.done("data"),
    });
  } else {
    hpMgr.ApiCmd("term-model/entry?" + uri, {
      callback: function (err, data) {
        ep.emit("data", {
          kind: "Term",
          model: data,
          id: "0",
          pid: "0",
          title: "",
          status: "1",
          weight: "0",
        });
      },
    });
  }
};

hpTerm.SetCommit = function () {
  var form = $("#hpm-termset"),
    alertid = "#hpm-node-alert";

  var req = {
    kind: "Term",
    id: parseInt(form.find("input[name=id]").val()),
    title: form.find("input[name=title]").val(),
    status: parseInt(form.find("input[name=status]").val()),
  };

  var model_type = form.find("input[name=model_type]").val();
  if ((model_type = "taxonomy")) {
    req.weight = parseInt(form.find("input[name=weight]").val());
    req.pid = parseInt(form.find("select[name=pid]").val());
  } else if ((model_type = "tag")) {
    //
  }

  //
  var uri =
    "modname=" +
    hpTerm.SpecActive() +
    "&modelid=" +
    hpTerm.SpecTermModelActive();

  hpMgr.ApiCmd("term/set?" + uri, {
    method: "POST",
    data: JSON.stringify(req),
    callback: function (err, data) {
      if (!data || data.kind != "Term") {
        return lynkui.alert.innerShow(alertid, "danger", data.error.message);
      }

      form.find("input[name=id]").val(data.id);

      lynkui.alert.innerShow(alertid, "success", "Successful operation");
      setTimeout(hpTerm.List, 500);
    },
  });
};

File: internal/datax/term.go
// Copyright 2015 Eryx <evorui аt gmаil dοt cοm>, All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package datax

import (
	"crypto/md5"
	"fmt"
	"io"
	"regexp"
	"strings"
	"time"

	"github.com/lessos/lessgo/types"

	"github.com/hooto/hpress/internal/config"
	"github.com/hooto/hpress/internal/hpapi"
	"github.com/hooto/hpress/internal/store"
)

var (
	spaceReg = regexp.MustCompile(" +")
)

func (q *QuerySet) TermCount() (int64, error) {

	table := hpapi.TermTableName(q.ModName, q.Table)

	fr := store.Data.NewFilter()
	fr.And("status", 1)

	return store.Data.Count(table, fr)
}

func (q *QuerySet) TermList() hpapi.TermList {

	rsp := hpapi.TermList{}

	model, err := config.SpecTermModel(q.ModName, q.Table)
	if err != nil {
		rsp.Error = &types.ErrorMeta{
			Code:    hpapi.ErrCodeBadArgument,
			Message: "Term Not Found",
		}
		return rsp
	}

	if model.Type == hpapi.TermTaxonomy {
		if tc, ok := termCMap[q.ModName+q.Table]; ok {
			return tc.ls
		}
	}

	// q.limit = 100
	table := hpapi.TermTableName(q.ModName, q.Table)

	qs := store.Data.NewQueryer().
		Select(q.cols).
		From(table).
		Offset(q.offset)

	if model.Type == hpapi.TermTag {
		qs.Order("updated desc")
	} else if model.Type == hpapi.TermTaxonomy {
		q.limit = 200
		qs.Order("weight desc")
	}

	qs.Limit(q.limit)

	qs.SetFilter(q.filter)

	rs := store.Data.Query(qs)
	if rs.Err() != nil {
		rsp.Error = &types.ErrorMeta{
			Code:    hpapi.ErrCodeInternalError,
			Message: "Can not pull database instance",
		}
		return rsp
	}

	for rs.Valid() {

		item := hpapi.Term{
			ID:      rs.Field("id").Uint32(),
			PID:     rs.Field("pid").Uint32(),
			Status:  rs.Field("status").Int16(),
			UserID:  rs.Field("userid").String(),
			Title:   rs.Field("title").String(),
			Created: rs.Field("created").Uint32(),
			Updated: rs.Field("updated").Uint32(),
		}

		switch model.Type {
		case hpapi.TermTag:
			item.UID = rs.Field("uid").String()
		case hpapi.TermTaxonomy:
			item.PID = rs.Field("pid").Uint32()
			item.Weight = rs.Field("weight").Int32()
		}

		rsp.Items = append(rsp.Items, item)

		rs.Next()
	}

	rsp.Model = model

	if q.Pager {
		num, _ := store.Data.Count(table, q.filter)
		rsp.Meta.TotalResults = uint64(num)
		rsp.Meta.StartIndex = uint64(q.offset)
		rsp.Meta.ItemsPerList = uint64(q.limit)
	}

	rsp.Kind = "TermList"

	if model.Type == hpapi.TermTaxonomy {

		tcm := &termCates{
			ls:  rsp,
			dps: map[uint32][]uint32{},
		}

		for _, termEntry := range tcm.ls.Items {
			tcm.dps[termEntry.ID] = termCateSubtree(&tcm.ls, []uint32{}, termEntry.ID)
		}

		termCMapMu.Lock()
		termCMap[q.ModName+q.Table] = tcm
		termCMapMu.Unlock()
	}

	// qryhash := q.Hash()

	// if model.CacheTTL > 0 && entry.Title != "" {
	// 	store.CacheSetJson(qryhash, rsp, model.CacheTTL)
	// }

	return rsp
}

func termCateSubtree(termls *hpapi.TermList, prs []uint32, pid uint32) []uint32 {

	if termInArray(prs, pid) {
		return prs
	}

	prs = append(prs, pid)

	for _, entry := range termls.Items {

		if entry.PID == pid {
			prs = termCateSubtree(termls, prs, entry.ID)
		}
	}

	return prs
}

func termInArray(arr []uint32, a uint32) bool {

	for _, ar := range arr {
		if ar == a {
			return true
		}
	}

	return false
}

func (q *QuerySet) TermEntry() hpapi.Term {

	var (
		rsp = hpapi.Term{}
		err error
	)

	rsp.Model, err = config.SpecTermModel(q.ModName, q.Table)
	if err != nil {
		rsp.Error = &types.ErrorMeta{
			Code:    hpapi.ErrCodeBadArgument,
			Message: "Term Not Found",
		}
		return rsp
	}

	table := hpapi.TermTableName(q.ModName, q.Table)

	qs := store.Data.NewQueryer().
		Select(q.cols).
		From(table).
		Order(q.order).
		Limit(1).
		Offset(q.offset)

	qs.SetFilter(q.filter)

	rs := store.Data.Query(qs)
	if rs.Err() != nil {
		rsp.Error = &types.ErrorMeta{
			Code:    hpapi.ErrCodeInternalError,
			Message: "Can not pull database instance",
		}
		return rsp
	}

	if rs.NotFound() {
		rsp.Error = &types.ErrorMeta{
			Code:    hpapi.ErrCodeBadArgument,
			Message: "Term Not Found",
		}
		return rsp
	}

	switch rsp.Model.Type {
	case hpapi.TermTaxonomy:
		rsp.PID = rs.Field("pid").Uint32()
		rsp.Weight = rs.Field("weight").Int32()
	case hpapi.TermTag:
		rsp.UID = rs.Field("uid").String()
	default:
		rsp.Error = &types.ErrorMeta{
			Code:    hpapi.ErrCodeInternalError,
			Message: "Server Error",
		}
		return rsp
	}

	rsp.ID = rs.Field("id").Uint32()
	rsp.PID = rs.Field("pid").Uint32()
	rsp.Status = rs.Field("status").Int16()
	rsp.UserID = rs.Field("userid").String()
	rsp.Title = rs.Field("title").String()
	rsp.Created = rs.Field("created").Uint32()
	rsp.Updated = rs.Field("updated").Uint32()

	rsp.Kind = "Term"

	// qryhash := q.Hash()

	// if rsp.Model.CacheTTL > 0 && entry.Title != "" {
	// 	store.CacheSetJson(qryhash, rsp, rsp.Model.CacheTTL)
	// }

	return rsp
}

type TermList hpapi.TermList

func (t *TermList) Index() string {

	if len(t.Items) < 1 {
		return ""
	}

	idxs := []string{}
	for _, v := range t.Items {
		idxs = append(idxs, fmt.Sprintf("%v", v.ID))
	}

	return strings.Join(idxs, ",")
}

func (t *TermList) Content() string {

	if len(t.Items) < 1 {
		return ""
	}

	ts := []string{}
	for _, v := range t.Items {
		ts = append(ts, v.Title)
	}

	return strings.Join(ts, ",")
}

func NodeTermQuery(modname string, model *hpapi.NodeModel, terms []hpapi.NodeTerm) []hpapi.NodeTerm {

	for _, modTerm := range model.Terms {

		for k, term := range terms {

			if modTerm.Meta.Name != term.Name {
				continue
			}

			switch modTerm.Type {

			case hpapi.TermTag:

				tags := strings.Split(term.Value, ",")

				for _, vtag := range tags {

					terms[k].Items = append(terms[k].Items, hpapi.Term{
						Title: vtag,
					})
				}

			case hpapi.TermTaxonomy:

				table := hpapi.TermTableName(modname, modTerm.Meta.Name)

				q := store.Data.NewQueryer().From(table)
				q.Limit(1)
				q.Where().And("id", term.Value)

				if rs := store.Data.Query(q); rs.OK() && !rs.NotFound() {

					terms[k].Items = append(terms[k].Items, hpapi.Term{
						ID:    rs.Field("id").Uint32(),
						Title: rs.Field("title").String(),
					})
				}
			}

			// terms[k].Type = modTerm.Type

			break
		}
	}

	return terms
}

func TermSync(modname, modelid, terms string) (TermList, error) {

	ls := TermList{}

	terms = spaceReg.ReplaceAllString(terms, " ")

	tars := strings.Split(terms, ",")

	ids := []interface{}{}

	for _, term := range tars {

		tag := hpapi.Term{
			Title: strings.TrimSpace(term),
		}

		if len(tag.Title) < 1 {
			continue
		}

		h := md5.New()
		io.WriteString(h, strings.ToLower(tag.Title))
		tag.UID = fmt.Sprintf("%x", h.Sum(nil))[:16]

		exist := false
		for _, prev := range ids {
			if prev.(string) == tag.UID {
				exist = true
				break
			}
		}
		if exist {
			continue
		}

		ls.Items = append(ls.Items, tag)

		ids = append(ids, tag.UID)
	}

	table := hpapi.TermTableName(modname, modelid)

	if len(ids) > 0 {

		q := store.Data.NewQueryer().From(table).Limit(int64(len(ids)))
		q.Where().And("uid.in", ids...)

		if rs := store.Data.Query(q); rs.OK() {

			for rs.Valid() {

				for tk, tv := range ls.Items {

					if rs.Field("uid").String() == tv.UID {

						ls.Items[tk].ID = rs.Field("id").Uint32()
						break
					}
				}

				rs.Next()
			}
		}
	}

	timenow := uint32(time.Now().Unix())

	for tk, tv := range ls.Items {

		if tv.ID > 0 {
			continue
		}

		if rs := store.Data.Insert(table, map[string]interface{}{
			"uid":     tv.UID,
			"title":   tv.Title,
			"userid":  "sysadmin",
			"status":  1,
			"created": timenow,
			"updated": timenow,
		}); rs.Err() == nil {

			// LastInsertId() is unsupported on pgsqlgo, so re-read the row by
			// its deterministic uid to get the server-assigned id.
			q := store.Data.NewQueryer().From(table).Limit(1)
			q.Where().And("uid", tv.UID)
			if rs2 := store.Data.Query(q); rs2.OK() && !rs2.NotFound() {
				ls.Items[tk].ID = rs2.Field("id").Uint32()
			}
		}
	}

	return ls, nil
}
